Code/Framework/AzCore/AzCore/DOM/DomVisitor.h

```diff
@@ -227,11 +227,18 @@ namespace AZ::DOM
 
         //! Helper method, constructs a failure \ref Result with the specified code.
         static Result VisitorFailure(VisitorErrorCode code);
-        //! Helper method, constructs a failure \ref Result with the specified code and supplemental info.
-        static Result VisitorFailure(VisitorErrorCode code, AZStd::string additionalInfo);
         //! Helper method, constructs a failure \ref Result with the specified error.
         static Result VisitorFailure(VisitorError error);
+
+        //! Helper method, constructs a failure \ref Result with the specified code and supplemental info specified by a format string
+        //! and its arguments.
+        template<class... TArgs>
+        static Result VisitorFailure(VisitorErrorCode code, TArgs... formatArgs)
+        {
+            return AZ::Failure(VisitorError(code, AZStd::string::format(formatArgs...)));
+        }
```

Comment:
I would have said the format string should be a separate argument to prevent any potential stack smashing exploits, but having a templated VisitorFailure function can result in many template instantiations that can be avoided, by having the caller just supply the string themselves.

Therefore I think this function should just accept an AZStd::string_view with the failure in it

```suggestion
        static Result VisitorFailure(VisitorErrorCode code, AZStd::string_view errorValue)
        {
            return AZ::Failure(VisitorError(code, errorValue));
        }
```

Reply:
I personally found it nice to have some templated helper for formatting errors but don't feel strongly about this - @AMZN-koppersr, this override was at your suggestion, do you have a strong preference? I can go back to taking `additionalInfo` as a string and call `format` manually if we'd like

Reply:
Answering that opens a can of worms around the dated approach that O3DE has to produce formatted issue messages. I would avoid making this too complex, but perhaps a solution that fits all needs is:
```
template<typename... TArgs>
static Result VisitorFailure(VisitorErrorCode code, const char* message, TArgs... formatArgs)
{
    if constexpr(sizeof...(formatArgs) == 0)
    {
        return AZ::Failure(VisitorError(code, message));
    }
    else
    {
        return AZ::Failure(VisitorError(code, AZStd::string::format(message, formatArgs...)))
    }
}

template<typename... TArgs>
static Result VisitorFailure(VisitorErrorCode code, const AZStd::string& message, TArgs... formatArgs)
{
    if constexpr(sizeof...(formatArgs) == 0)
    {
        return AZ::Failure(VisitorError(code, message));
    }
    else
    {
        return AZ::Failure(VisitorError(code, AZStd::string::format(message.c_str(), formatArgs...)))
    }
}

template<typename... TArgs>
static Result VisitorFailure(VisitorErrorCode code, AZStd::string&& message, TArgs... formatArgs)
{
    if constexpr(sizeof...(formatArgs) == 0)
    {
        return AZ::Failure(VisitorError(code, AZStd::move(message)));
    }
    else
    {
        return AZ::Failure(VisitorError(code, AZStd::string::format(message.c_str(), formatArgs...)))
    }
}

static Result VisitorFailure(VisitorErrorCode code, AZStd::string_view message)
{
    return AZ::Failure(VisitorError(code, message));
}
```
(Probably needs some more tweaking and screams for a more generic solution, which is that can of worms I mentioned.)

Reply:
My major concern with this code is that the template functions don't make this anymore safer than it just producing the formatted string outside of the method.

Invoking VisitorFailure with the wrong type of arguments still isn't caught, because the formatted is still happening in the variadic function.

```
int failureCount = 7;
VisitorFailure(VisitorErrorCode::UnsupportedOperation, "This operation has failed %s times", failureCount);
```

That snippet I just posted would compile and result in a format error the same way of using `AZStd::string::format` directly would.

Reply:
I ended up just getting rid of the helper and using `AZStd::string::format` instead - more verbose, but no chance of accidentally misusing it, so I think it's for the best.